govcd/vdc_group.go

```diff
@@ -0,0 +1,478 @@
+}
+
+// Delete deletes VDC group
+func (vdcGroup *VdcGroup) Delete() error {
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointVdcGroups
+	minimumApiVersion, err := vdcGroup.client.checkOpenApiEndpointCompatibility(endpoint)
+	if err != nil {
+		return err
+	}
+
+	if vdcGroup.VdcGroup.Id == "" {
+		return fmt.Errorf("cannot delete VDC group without id")
+	}
+
+	urlRef, err := vdcGroup.client.OpenApiBuildEndpoint(endpoint, vdcGroup.VdcGroup.Id)
+	if err != nil {
+		return err
+	}
+
+	err = vdcGroup.client.OpenApiDeleteItem(minimumApiVersion, urlRef, nil, nil)
+
+	if err != nil {
+		return fmt.Errorf("error deleting VDC group: %s", err)
+	}
+
+	return nil
+}
+
+// GetAllVdcGroups retrieves all VDC groups. Query parameters can be supplied to perform additional filtering
+func (adminOrg *AdminOrg) GetAllVdcGroups(queryParameters url.Values) ([]*VdcGroup, error) {
+	tenantContext, err := adminOrg.getTenantContext()
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointVdcGroups
+	minimumApiVersion, err := adminOrg.client.checkOpenApiEndpointCompatibility(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := adminOrg.client.OpenApiBuildEndpoint(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	responses := []*types.VdcGroup{}
+	err = adminOrg.client.OpenApiGetAllItems(minimumApiVersion, urlRef, queryParameters, &responses, getTenantContextHeader(tenantContext))
+	if err != nil {
+		return nil, err
+	}
+
+	var wrappedVdcGroups []*VdcGroup
+	for _, response := range responses {
+		urlRef, err := adminOrg.client.OpenApiBuildEndpoint(endpoint, response.Id)
+		if err != nil {
+			return nil, err
+		}
+		wrappedVdcGroup := &VdcGroup{
+			VdcGroup: response,
+			client:   adminOrg.client,
+			Href:     urlRef.String(),
+			parent:   adminOrg,
+		}
+		wrappedVdcGroups = append(wrappedVdcGroups, wrappedVdcGroup)
+	}
+
+	return wrappedVdcGroups, nil
+}
+
+// GetVdcGroupByName retrieves VDC group by given name
+// When the name contains commas, semicolons or asterisks, the encoding is rejected by the API in VCD 10.2 version.
+// For this reason, when one or more commas, semicolons or asterisks are present we run the search brute force,
+// by fetching all VDC groups and comparing the names. Yet, this not needed anymore in VCD 10.3 version.
+// Also, url.QueryEscape as well as url.Values.Encode() both encode the space as a + character. So we use
+// search brute force too. Reference to issue:
+// https://github.com/golang/go/issues/4013
+// https://github.com/czos/goamz/pull/11/files
+func (adminOrg *AdminOrg) GetVdcGroupByName(name string) (*VdcGroup, error) {
+	slowSearch, params, err := isShouldDoSlowSearch("name", name, adminOrg.client)
+	if err != nil {
+		return nil, err
+	}
+
+	vdcGroups, err := adminOrg.GetAllVdcGroups(params)
+	if err != nil {
+		return nil, err
+	}
+	if len(vdcGroups) == 0 {
+		return nil, ErrorEntityNotFound
+	}
+
+	if slowSearch {
```

Comment:
Behavior will be a bit different for the ones that use `slowSearch`and not.

In the regular case it will
* return ErrorEntityNotFound when nothing is found
* return error of `more than one VDC group found with name '%s'` when >1 VDC groups are found

In the slow search case it will
* return ErrorEntityNotFound when nothing is found
* return *first* result if more than one is found

Reply:
changed to have same behavior